src/recovery.rs

```diff
@@ -97,6 +658,13 @@ pub(crate) struct PartitionMeta(PartitionIndex, PartitionCount);
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub(crate) struct ExecutionNumber(u64);
 
+impl ExecutionNumber {
+    pub(crate) fn next(mut self) -> Self {
```

Comment:
I used `mut self` so that the function consumes the struct to make sure the old execution number is not used again by mistake, but I'm ok with returning a separate instance too, we still need to be careful either way.